File: sdk/verifiers/python/src/pipelock_aarp_verify/appraise.py
# Copyright 2026 Josh Waldrep
# SPDX-License-Identifier: Apache-2.0

"""Envelope appraisal, claim classification, and the comparable output.

Ports ``internal/aarp/verify.py``, ``appraise.py``, and ``comparable.py``. The
appraisal never reports "trusted" or "safe"; it reports which claims the verifier
could cryptographically confirm, grouped by axis, plus the fixed does_not_assert
list. Per-signature problems are reported per signature, never as an envelope
rejection, so one bad parallel signature cannot mask a good one.
"""

from __future__ import annotations

import base64
import binascii
from dataclasses import dataclass, field
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .canonical import canonicalize
from .envelope import Envelope, ProtectedHeader, Signature, signing_input
from .suite import (
    CANON_ID,
    IMPLEMENTED_ALGS,
    KEY_TYPE_FOR_ALG,
    KNOWN_SIGNER_ROLES,
    PROFILE,
    MalformedSuiteError,
    UnknownCriticalExtensionError,
    UnknownSuiteError,
    check_critical_extensions,
)

ED25519_PUBLIC_KEY_SIZE = 32
ED25519_SIGNATURE_SIZE = 64

# Axis names grouping verified claims by the kind of proof they rest on.
AXIS_IDENTITY = "identity"
AXIS_INTEGRITY = "integrity"
AXIS_FRESHNESS = "freshness"

# Verified-claim names.
CLAIM_ASSERTION_SIGNATURE_VALID = "assertion_signature_valid"
CLAIM_MEDIATOR_KEY_PINNED = "mediator_key_pinned"
CLAIM_CHAIN_LINK_PRESENT = "chain_link_present"
CLAIM_WORKLOAD_IDENTITY_VERIFIED = "workload_identity_verified"
CLAIM_X509_SVID_BOUND = "x509_svid_bound"
CLAIM_SVID_VALID_AT_ACTION_TIME = "svid_valid_at_action_time"

# Per-signature status enum.
SIG_VERIFIED = "verified"
SIG_FAILED = "failed"
SIG_UNKNOWN_KEY = "unknown_key"
SIG_UNIMPLEMENTED = "unimplemented"
SIG_UNKNOWN_SUITE = "unknown_suite"
SIG_MALFORMED = "malformed"

# The fixed set of properties an AARP appraisal never asserts.
DOES_NOT_ASSERT = [
    "efficacy",
    "absence_of_bypass",
    "complete_mediation",
    "policy_correctness",
    "action_safety",
]

# Producer claim -> required verified claims (all must be present to confirm).
# None means structurally claim-only in v0.1 (never verifiable).
CLAIM_VERIFIED_BY: dict[str, list[str] | None] = {
    "mediated": [CLAIM_MEDIATOR_KEY_PINNED],
    "complete-mediation": None,
    "complete_mediation": None,
    CLAIM_WORKLOAD_IDENTITY_VERIFIED: [CLAIM_WORKLOAD_IDENTITY_VERIFIED],
    CLAIM_X509_SVID_BOUND: [CLAIM_X509_SVID_BOUND],
    CLAIM_SVID_VALID_AT_ACTION_TIME: [CLAIM_SVID_VALID_AT_ACTION_TIME],
    "transparency_inclusion": None,
}


@dataclass
class TrustEntry:
    """Binds a signing key id to an authority namespace, never a bare key."""

    mediator_id: str = ""
    role: str = ""
    trust_domain: str = ""


@dataclass
class VerifyOptions:
    """The verifier's pinned trust. Never anything fetched live."""

    trusted_keys: dict[str, bytes] = field(default_factory=dict)
    trust: dict[str, TrustEntry] = field(default_factory=dict)


@dataclass
class SignatureResult:
    """The appraisal of one parallel signature."""

    key_id: str
    alg: str
    signer_role: str
    status: str
    reason: str = ""


@dataclass
class Appraisal:
    """The AARP verifier result."""

    profile: str = PROFILE
    assertion_signed: bool = False
    signatures: list[SignatureResult] = field(default_factory=list)
    assurance_claimed: list[str] = field(default_factory=list)
    verified_claims: list[str] = field(default_factory=list)
    claimed_unverified: list[str] = field(default_factory=list)
    axes: dict[str, list[str]] = field(default_factory=dict)
    does_not_assert: list[str] = field(default_factory=lambda: list(DOES_NOT_ASSERT))
    warnings: list[str] = field(default_factory=list)

    def add_verified(self, claim: str, axis: str) -> None:
        self.verified_claims.append(claim)
        self.axes.setdefault(axis, []).append(claim)


class EnvelopeFatalError(Exception):
    """The envelope is fatal to appraise (schema/profile/canon/crit-ext)."""


def _validate_structure(env: Envelope) -> None:
    """Envelope-fatal schema checks that must hold before per-signature appraisal.

    Only fields inside the SIGNED payload are envelope-fatal here: the payload
    parts (which include the top-level profile and the envelope-level
    critical-extension list) and a non-empty signature set. Per-signature suite
    fields (a signature's protected profile, canon, alg, key trust, and its own
    critical-extension list) are deliberately NOT checked here — they are
    per-signature outcomes appraised in ``_appraise_signature``. The signatures
    array is not itself signed, so a man-in-the-middle can append a junk
    signature; making a bad protected header envelope-fatal would let that append
    deny a legitimately-signed envelope.
    """
    env.validate_payload_parts()
    if len(env.signatures) == 0:
        raise EnvelopeFatalError("envelope has no signatures")


def verify(env: Envelope, opts: VerifyOptions) -> Appraisal:
    """Appraise an envelope. Raises EnvelopeFatalError for envelope-fatal cases.

    Envelope-fatal means a violation in the SIGNED payload: a schema error, the
    top-level profile mismatch, an unsafe number, an unknown AARP field, a bad
    typed-string grammar, or an unknown/malformed envelope-level critical
    extension. Per-signature suite problems (a signature's protected
    profile/canon, an unknown or unimplemented alg, its own critical extensions,
    an untrusted key, or an invalid signature) are reported per signature and
    never raise here, so one bad parallel signature cannot mask a good one.
    """
    ap = _appraise_core(env, opts)
    _classify_claims(ap)
    return ap


def _appraise_core(env: Envelope, opts: VerifyOptions) -> Appraisal:
    _validate_structure(env)
    digest = env.payload_digest()

    ap = Appraisal()
    ap.assurance_claimed = list(env.assertion.claimed)

    verified: list[tuple[str, str]] = []  # (key_id, role)
    for s in env.signatures:
        res, ok = _appraise_signature(s, digest, opts)
        ap.signatures.append(res)
        if ok:
            verified.append((s.protected.key_id, s.protected.signer_role))

    if verified:
        ap.assertion_signed = True
        ap.add_verified(CLAIM_ASSERTION_SIGNATURE_VALID, AXIS_INTEGRITY)
        if _mediator_key_pinned(env.assertion, verified, opts.trust):
            ap.add_verified(CLAIM_MEDIATOR_KEY_PINNED, AXIS_IDENTITY)
        if env.chain is not None:
            ap.add_verified(CLAIM_CHAIN_LINK_PRESENT, AXIS_INTEGRITY)
    else:
        ap.warnings.append(
            "no signature verified under a trusted key; all assurance claims "
            "are untrusted input"
        )
    return ap


def _appraise_signature(
    s: Signature, digest: str, opts: VerifyOptions
) -> tuple[SignatureResult, bool]:
    """Verify one parallel signature. Never falls back to a different suite."""
    h: ProtectedHeader = s.protected
    res = SignatureResult(
        key_id=h.key_id, alg=h.alg, signer_role=h.signer_role, status=""
    )

    # Per-signature suite identity. A wrong profile/canon or an unknown critical
    # extension in THIS signature's protected header makes only this signature
    # unverifiable — it never rejects the envelope, so an appended junk signature
    # cannot deny a verifiable sibling. (The signed top-level profile and
    # crit_ext are checked envelope-fatal in _validate_structure.)
    if h.profile != PROFILE:
        res.status = SIG_UNKNOWN_SUITE
        res.reason = f"profile {h.profile!r} != {PROFILE!r}"
        return res, False
    if h.canon != CANON_ID:
        res.status = SIG_UNKNOWN_SUITE
        res.reason = f"canon {h.canon!r} != {CANON_ID!r}"
        return res, False
    try:
        check_critical_extensions(h.crit)
    except UnknownCriticalExtensionError as exc:
        res.status, res.reason = SIG_UNKNOWN_SUITE, str(exc)
        return res, False
    except MalformedSuiteError as exc:
        res.status, res.reason = SIG_MALFORMED, str(exc)
        return res, False

    if h.key_id == "":
        res.status, res.reason = SIG_MALFORMED, "empty key_id"
        return res, False
    if h.signer_role not in KNOWN_SIGNER_ROLES:
        res.status, res.reason = SIG_MALFORMED, "unknown signer_role"
        return res, False
    want_key_type = KEY_TYPE_FOR_ALG.get(h.alg)
    if want_key_type is None:
        res.status, res.reason = (
            SIG_UNKNOWN_SUITE,
            "unrecognized algorithm; no fallback",
        )
        return res, False
    if h.key_type != want_key_type:
        res.status = SIG_MALFORMED
        res.reason = f"key_type {h.key_type!r} != {want_key_type!r} required by alg"
        return res, False
    if h.alg not in IMPLEMENTED_ALGS:
        res.status, res.reason = (
            SIG_UNIMPLEMENTED,
            "recognized suite, verifier not yet built",
        )
        return res, False

    # Implemented suite: Ed25519.
    pub = opts.trusted_keys.get(h.key_id)
    if pub is None:
        res.status, res.reason = SIG_UNKNOWN_KEY, "key_id not in trusted set"
        return res, False
    if len(pub) != ED25519_PUBLIC_KEY_SIZE:
        res.status, res.reason = SIG_MALFORMED, "trusted key has wrong size"
        return res, False
    try:
        message = signing_input(digest, h)
    except Exception as exc:  # noqa: BLE001 - mirror Go's malformed-on-error
        res.status, res.reason = SIG_MALFORMED, f"signing input: {exc}"
        return res, False
    raw = _decode_sig_wire(h.alg, s.sig)
    if raw is None:
        res.status, res.reason = SIG_MALFORMED, "signature wire malformed"
        return res, False
    if len(raw) != ED25519_SIGNATURE_SIZE:
        res.status, res.reason = (
            SIG_FAILED,
            "signature does not verify over canonical bytes",
        )
        return res, False
    try:
        Ed25519PublicKey.from_public_bytes(pub).verify(raw, message)
    except InvalidSignature:
        res.status, res.reason = (
            SIG_FAILED,
            "signature does not verify over canonical bytes",
        )
        return res, False
    res.status = SIG_VERIFIED
    return res, True


def _decode_sig_wire(alg: str, wire: str) -> bytes | None:
    """Split an "<alg>:<base64-std>" signature into raw bytes, or None if bad."""
    prefix = alg + ":"
    if not wire.startswith(prefix):
        return None
    try:
        return base64.standard_b64decode(wire[len(prefix) :])
    except (binascii.Error, ValueError):
        return None


def _mediator_key_pinned(
    assertion: Any, verified: list[tuple[str, str]], trust: dict[str, TrustEntry]
) -> bool:
    """Whether a verifying signature is trust-bound to the asserted mediator."""
    for key_id, role in verified:
        entry = trust.get(key_id)
        if entry is None:
            continue
        if entry.mediator_id != assertion.mediator_id:
            continue
        if entry.trust_domain != "" and entry.trust_domain != assertion.trust_domain:
            continue
        if entry.role != "" and entry.role != role:
            continue
        return True
    return False


def _classify_claims(ap: Appraisal) -> None:
    """Fill claimed_unverified from producer claims the verifier did not confirm."""
    verified = set(ap.verified_claims)
    seen: set[str] = set()
    for claimed in ap.assurance_claimed:
        if claimed in seen:
            continue
        seen.add(claimed)
        if claimed not in CLAIM_VERIFIED_BY:
            ap.claimed_unverified.append(claimed)
            ap.warnings.append(
                "unknown assurance claim reported claim-only: " + claimed
            )
            continue
        required = CLAIM_VERIFIED_BY[claimed]
        if not required:  # None or empty list -> never verifiable
            ap.claimed_unverified.append(claimed)
            continue
        if all(r in verified for r in required):
            continue
        ap.claimed_unverified.append(claimed)


def _sorted_unique(items: list[str]) -> list[str]:
    return sorted(set(items))


def comparable_appraisal(ap: Appraisal) -> bytes:
    """Project an Appraisal onto the cross-language comparison surface (JCS bytes).

    EXCLUDES warnings, per-signature reason text, and assurance_claimed. The
    signatures array preserves envelope order; all other arrays are sorted and
    deduplicated.
    """
    sigs: list[Any] = [
        {
            "alg": s.alg,
            "key_id": s.key_id,
            "signer_role": s.signer_role,
            "status": s.status,
        }
        for s in ap.signatures
    ]
    axes: dict[str, Any] = {}
    for axis, claims in ap.axes.items():
        if not claims:
            continue
        axes[axis] = _sorted_unique(claims)
    obj = {
        "profile": ap.profile,
        "assertion_signed": ap.assertion_signed,
        "signatures": sigs,
        "verified_claims": _sorted_unique(ap.verified_claims),
        "claimed_unverified": _sorted_unique(ap.claimed_unverified),
        "axes": axes,
        "does_not_assert": _sorted_unique(ap.does_not_assert),
    }
    return canonicalize(obj)


# Bind so a caller can treat suite/crit-ext failures as fatal uniformly.
FATAL_EXCEPTIONS = (
    EnvelopeFatalError,
    UnknownSuiteError,
    UnknownCriticalExtensionError,
)

File: sdk/verifiers/ts/src/aarp/appraise.ts
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

// Appraisal vocabulary, claim classification, signature verification, and the
// comparable output projection. Ported from internal/aarp/appraise.go,
// verify.go, and comparable.go.

import { createPublicKey, verify as cryptoVerify } from "node:crypto";
import { canonicalize } from "./canonical.js";
import {
  Assertion,
  Envelope,
  Signature,
  payloadDigest,
  signingInput,
  validateStructure,
} from "./envelope.js";
import {
  ALG_ED25519,
  CANON_ID,
  MalformedCritError,
  PROFILE,
  UnknownCritError,
  checkCriticalExtensions,
  implementedAlgs,
  keyTypeForAlg,
  knownSignerRoles,
} from "./suite.js";

// Axis names group verified claims by the kind of proof they rest on.
export const AXIS_IDENTITY = "identity";
export const AXIS_INTEGRITY = "integrity";

// Verified-claim names.
export const CLAIM_ASSERTION_SIGNATURE_VALID = "assertion_signature_valid";
export const CLAIM_MEDIATOR_KEY_PINNED = "mediator_key_pinned";
// A signed, well-formed Rung-1 chain link is present on the envelope (its
// position is authenticated by the verified signature). Mirrors Go's
// ClaimChainLinkPresent = "chain_link_present". This is the single-envelope
// appraisal claim; the cross-envelope stream linkage ("chain_linked") is a
// separate concept reported by verifyChain/comparableChain.
export const CLAIM_CHAIN_LINK_PRESENT = "chain_link_present";

// docsNotAsserted is the fixed set of properties an AARP appraisal never asserts.
const docsNotAsserted = [
  "efficacy",
  "absence_of_bypass",
  "complete_mediation",
  "policy_correctness",
  "action_safety",
];

// SignatureStatus is the per-signature appraisal outcome.
export type SignatureStatus =
  | "verified"
  | "failed"
  | "unknown_key"
  | "unimplemented"
  | "unknown_suite"
  | "malformed";

export interface SignatureResult {
  key_id: string;
  alg: string;
  signer_role: string;
  status: SignatureStatus;
  reason?: string;
}

export interface Appraisal {
  profile: string;
  assertion_signed: boolean;
  signatures: SignatureResult[];
  assurance_claimed: string[];
  verified_claims: string[];
  claimed_unverified: string[];
  axes: Record<string, string[]>;
  does_not_assert: string[];
  warnings: string[];
}

export interface TrustEntry {
  mediator_id: string;
  role?: string;
  trust_domain?: string;
}

export interface VerifyOptions {
  // trustedKeys maps a key id to its raw 32-byte Ed25519 public key.
  trustedKeys: Map<string, Uint8Array>;
  // trust maps a key id to its authority-namespace binding.
  trust: Map<string, TrustEntry>;
}

// claimVerifiedBy maps each producer claim to the verified claims required for it
// to count as confirmed. An empty list means structurally claim-only.
const claimVerifiedBy: Record<string, string[]> = {
  mediated: [CLAIM_MEDIATOR_KEY_PINNED],
  "complete-mediation": [],
  complete_mediation: [],
  transparency_inclusion: [],
};

interface VerifiedSigner {
  keyID: string;
  role: string;
}

// ed25519 SPKI DER prefix for a raw 32-byte public key.
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

function ed25519PublicKey(raw: Uint8Array) {
  const der = Buffer.concat([SPKI_PREFIX, Buffer.from(raw)]);
  return createPublicKey({ key: der, format: "der", type: "spki" });
}

function decodeSigWire(alg: string, wire: string): Buffer | null {
  const prefix = `${alg}:`;
  if (!wire.startsWith(prefix)) return null;
  const b64 = wire.slice(prefix.length);
  // Strict standard-base64 round-trip check: Node's base64 decoder is lenient,
  // so re-encode and compare to reject malformed input the way Go does.
  const raw = Buffer.from(b64, "base64");
  if (raw.toString("base64") !== b64) return null;
  return raw;
}

// appraiseSignature verifies one parallel signature. It never falls back.
function appraiseSignature(
  s: Signature,
  digest: string,
  opts: VerifyOptions,
): { result: SignatureResult; ok: boolean } {
  const res: SignatureResult = {
    key_id: s.protected.key_id,
    alg: s.protected.alg,
    signer_role: s.protected.signer_role,
    status: "malformed",
  };

  // Per-signature suite identity. A wrong profile/canon or an unknown critical
  // extension in THIS signature's protected header makes only this signature
  // unverifiable - it never rejects the envelope, so an appended junk signature
  // cannot deny a verifiable sibling. (The signed top-level profile and
  // envelope-level crit_ext are checked envelope-fatal in validateStructure.)
  if (s.protected.profile !== PROFILE) {
    res.status = "unknown_suite";
    res.reason = `profile ${JSON.stringify(s.protected.profile)} != ${JSON.stringify(PROFILE)}`;
    return { result: res, ok: false };
  }
  if (s.protected.canon !== CANON_ID) {
    res.status = "unknown_suite";
    res.reason = `canon ${JSON.stringify(s.protected.canon)} != ${JSON.stringify(CANON_ID)}`;
    return { result: res, ok: false };
  }
  try {
    checkCriticalExtensions(s.protected.crit);
  } catch (err) {
    if (err instanceof UnknownCritError) {
      res.status = "unknown_suite";
      res.reason = err.message;
    } else if (err instanceof MalformedCritError) {
      res.status = "malformed";
      res.reason = err.message;
    } else {
      throw err;
    }
    return { result: res, ok: false };
  }

  if (s.protected.key_id === "") {
    res.status = "malformed";
    res.reason = "empty key_id";
    return { result: res, ok: false };
  }
  if (!knownSignerRoles[s.protected.signer_role]) {
    res.status = "malformed";
    res.reason = "unknown signer_role";
    return { result: res, ok: false };
  }
  const wantKeyType = keyTypeForAlg[s.protected.alg];
  if (wantKeyType === undefined) {
    res.status = "unknown_suite";
    res.reason = "unrecognized algorithm; no fallback";
    return { result: res, ok: false };
  }
  if (s.protected.key_type !== wantKeyType) {
    res.status = "malformed";
    res.reason = `key_type ${JSON.stringify(s.protected.key_type)} != ${JSON.stringify(wantKeyType)} required by alg`;
    return { result: res, ok: false };
  }
  if (!implementedAlgs[s.protected.alg]) {
    res.status = "unimplemented";
    res.reason = "recognized suite, verifier not yet built";
    return { result: res, ok: false };
  }

  // Implemented suite: Ed25519.
  const pub = opts.trustedKeys.get(s.protected.key_id);
  if (pub === undefined) {
    res.status = "unknown_key";
    res.reason = "key_id not in trusted set";
    return { result: res, ok: false };
  }
  if (pub.length !== 32) {
    res.status = "malformed";
    res.reason = "trusted key has wrong size";
    return { result: res, ok: false };
  }
  const raw = decodeSigWire(s.protected.alg, s.sig);
  if (raw === null) {
    res.status = "malformed";
    res.reason = "signature wire malformed";
    return { result: res, ok: false };
  }
  if (raw.length !== 64) {
    res.status = "failed";
    res.reason = "signature does not verify over canonical bytes";
    return { result: res, ok: false };
  }
  const input = signingInput(digest, s.protected);
  let valid = false;
  try {
    valid = cryptoVerify(null, input, ed25519PublicKey(pub), raw);
  } catch {
    valid = false;
  }
  if (!valid) {
    res.status = "failed";
    res.reason = "signature does not verify over canonical bytes";
    return { result: res, ok: false };
  }
  res.status = "verified";
  return { result: res, ok: true };
}

// mediatorKeyPinned reports whether any verifying signature is bound by a trust
// entry to the asserted mediator identity (role/domain-scoped).
function mediatorKeyPinned(
  a: Assertion,
  verified: VerifiedSigner[],
  trust: Map<string, TrustEntry>,
): boolean {
  for (const vs of verified) {
    const entry = trust.get(vs.keyID);
    if (entry === undefined) continue;
    if (entry.mediator_id !== a.mediator_id) continue;
    if (
      entry.trust_domain !== undefined &&
      entry.trust_domain !== "" &&
      entry.trust_domain !== (a.trust_domain ?? "")
    ) {
      continue;
    }
    if (entry.role !== undefined && entry.role !== "" && entry.role !== vs.role) continue;
    return true;
  }
  return false;
}

function newAppraisal(): Appraisal {
  return {
    profile: "aarp/v0.1",
    assertion_signed: false,
    signatures: [],
    assurance_claimed: [],
    verified_claims: [],
    claimed_unverified: [],
    axes: {},
    does_not_assert: [...docsNotAsserted],
    warnings: [],
  };
}

function addVerified(ap: Appraisal, claim: string, axis: string): void {
  ap.verified_claims.push(claim);
  if (ap.axes[axis] === undefined) ap.axes[axis] = [];
  (ap.axes[axis] as string[]).push(claim);
}

function classifyClaims(ap: Appraisal): void {
  const verified = new Set(ap.verified_claims);
  const seenClaim = new Set<string>();
  for (const claimed of ap.assurance_claimed) {
    if (seenClaim.has(claimed)) continue;
    seenClaim.add(claimed);
    const required = claimVerifiedBy[claimed];
    if (required === undefined) {
      ap.claimed_unverified.push(claimed);
      ap.warnings.push(`unknown assurance claim reported claim-only: ${claimed}`);
      continue;
    }
    if (required.length === 0) {
      ap.claimed_unverified.push(claimed);
      continue;
    }
    if (required.every((r) => verified.has(r))) continue;
    ap.claimed_unverified.push(claimed);
  }
}

// verify appraises an AARP envelope. It throws an EnvelopeFatal-class error only
// for envelope-fatal conditions; per-signature problems are reported in the
// appraisal, never as a hard rejection.
export function verify(e: Envelope, opts: VerifyOptions): Appraisal {
  validateStructure(e);
  const digest = payloadDigest(e);

  const ap = newAppraisal();
  ap.profile = e.profile;
  ap.assurance_claimed = [...e.assertion.claimed];

  const verified: VerifiedSigner[] = [];
  for (const s of e.signatures) {
    const { result, ok } = appraiseSignature(s, digest, opts);
    ap.signatures.push(result);
    if (ok) {
      verified.push({ keyID: s.protected.key_id, role: s.protected.signer_role });
    }
  }

  if (verified.length > 0) {
    ap.assertion_signed = true;
    addVerified(ap, CLAIM_ASSERTION_SIGNATURE_VALID, AXIS_INTEGRITY);
    if (mediatorKeyPinned(e.assertion, verified, opts.trust)) {
      addVerified(ap, CLAIM_MEDIATOR_KEY_PINNED, AXIS_IDENTITY);
    }
    if (e.chain !== undefined) {
      addVerified(ap, CLAIM_CHAIN_LINK_PRESENT, AXIS_INTEGRITY);
    }
  } else {
    ap.warnings.push(
      "no signature verified under a trusted key; all assurance claims are untrusted input",
    );
  }

  classifyClaims(ap);
  return ap;
}

// sortedUnique returns a sorted, de-duplicated copy of in.
function sortedUnique(input: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const s of input) {
    if (seen.has(s)) continue;
    seen.add(s);
    out.push(s);
  }
  out.sort();
  return out;
}

// comparableAppraisal projects an appraisal onto the cross-language comparison
// surface and returns its JCS-canonical bytes. Excludes warnings, per-signature
// reason, and assurance_claimed.
export function comparableAppraisal(ap: Appraisal): string {
  const sigs = ap.signatures.map((s) => ({
    alg: s.alg,
    key_id: s.key_id,
    signer_role: s.signer_role,
    status: s.status,
  }));

  const axes: Record<string, unknown> = {};
  for (const [axis, claims] of Object.entries(ap.axes)) {
    if (claims.length === 0) continue;
    axes[axis] = sortedUnique(claims);
  }

  const obj = {
    profile: ap.profile,
    assertion_signed: ap.assertion_signed,
    signatures: sigs,
    verified_claims: sortedUnique(ap.verified_claims),
    claimed_unverified: sortedUnique(ap.claimed_unverified),
    axes,
    does_not_assert: sortedUnique(ap.does_not_assert),
  };
  return canonicalize(obj);
}
